examples/README.md

```diff
@@ -216,3 +221,34 @@ spec:
 
 If you access the pod you can check that your data is avaialable at
 `/var/lib/www/html`
+
+## Cluster ID based configuration
+
+Before creating a storage class that uses the option `clusterID` to refer to a
+Ceph cluster,
+
+**NOTE**: Substitute the output of `ceph fsid` instead of `<cluster-fsid>` in
+    the mentioned template YAML files, and also the Ceph admin ID and
+    credentials in their respective options. Further, update options like
+    `monitors` and `pools` in the respective YAML files to contain the
+    appropriate information.
+
+Create the following config maps and secrets
+
+* `kubectl create -f ./rbd/template-ceph-cluster-ID-provisioner-secret.yaml`
+* `kubectl create -f ./rbd/template-ceph-cluster-ID-publish-secret.yaml`
+* `kubectl create -f ./rbd/template-ceph-cluster-ID-config.yaml`
+
+Modify the deployed CSI pods to additionally pass in the config maps and
+secrets as volumes,
+
+* `kubectl patch daemonset csi-rbdplugin --patch "$(cat ./rbd/template-csi-rbdplugin-patch.yaml)"`
```

Comment:
Let csi plugins fetch Secrets/Configmaps as they do now. We should not inject secret into the the csi driver pods.

Reply:
1. If we retain or use secrets in the StorageClass, then we would end up with the [cons](https://github.com/ceph/ceph-csi/blob/7c2f755b6b2420f33c60c09d90c30e90ee7d6564/docs/design/proposals/stateless-csi-plugin.md#cons-alternative-2) as in alternative-2 in the proposal. (IOW, no listing and GetCapacity support is feasible)

2. Further the k8s CSI development guide states in the first [para](https://kubernetes-csi.github.io/docs/secrets-and-credentials.html#csi-driver-secrets) of its secret management, that driver level secrets can be injected into the driver,

quoted here:
```
Some drivers may require a secret in order to issue operations against a backend
(a service account, for example). If this secret is required at the "per driver"
granularity (and not different "per CSI operation" or "per volume"), the secret may
be injected in to CSI driver pods via standard Kubernetes secret distribution
mechanisms.
```

So I assume we are not breaking any CSI implementation rules as far as I can see.

3. We need to edit the pod manifest anyway to add per cluster config map, when we add a cluster using which the CSI driver can provision volumes. Hence, from a usability standpoint, injecting the secrets would happen at the same time.

4. We also default to checking the passed in credentials [first](https://github.com/ceph/ceph-csi/pull/241/files#diff-8e27516879efe971bf4755fadf585e06R94) before accessing the injected credential store, so that if needed the StorageClass can override the injected secrets and the driver would use the same.

Given the above, should we really mandate that the CSI driver not be injected with the secrets? What am I missing?

Reply:
with currently proposed approach if we want to add new cluster info we need to patch the plugin again

I think as rbd provisioner has the permission to fetch both the secret and the config map
can we fetch the configuration on the fly, 
suppose we list the config map with a name `rbd-cluster-config-*`, I think we can watch for the new config map added or changes and update local cache based on that? (if local cache is not required we can fetch whenever its required)
same can be done for secret also as we are having clusterid

@ShyamsundarR @rootfs @gman0 let me know your thoughts on this

Reply:
@Madhu-1 are you referring to a scheme where the config map and secrets are not mapped in as volumes, but that we could fetch them programmatically? IOW, [like so](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#watch-configmap-v1-core)

This seems kube specific, but also a useful approach to prevent pod restarts. As CSI is supposed to be CO independent, I would propose to retain the file mapped approach, and later add in the kube specific watch/fetch of config maps and secrets as an enhancement when running on kube based orchestration environments.

Reply:
I have a few follow-up questions, 
* are we planning to support ceph-csi driver for multiple CO?
* current readme is  all about k8s (need to change to multiple CO specific)
*  if we want to support multiple CO are we going to test driver with all CO (which support CSI specification)? 
* am not sure config map or secrets behave in the same  way in all CO

am towards fetching configmap and secret  from which supports multiple cluster on the fly without patching the plugin each time we add a new cluster config (avoids  pathcing secrets,configmap etc)

am sure other CO will have this feature, but we need to figure it out if we  want to support multiple CO

Reply:
> am towards fetching configmap and secret from which supports multiple cluster on the fly without patching the plugin each time we add a new cluster config (avoids pathcing secrets,configmap etc)

As stated earlier, I agree, it is more elegant than having to restart the pods on config map changes. But, I would like the keep k8s specifics out of the code, or pluggable.

Are you stating we make the change right away in this PR, or as one of the next steps? I prefer the latter, as getting to a volume-ID that can be mapped back to a Ceph image without the config map that stores the mapping is IMO more key than getting to an auto discoverable config (which is, to repeat, a better place to be than reloading the pods).
 
> am sure other CO will have this feature, but we need to figure it out if we want to support multiple CO

The intent of CSI has been to keep plugins CO independent. If code can retain the same, and not have to worry about it when the next CO that needs to be supported, comes into play, it would be cleaner IMO. In this vein, to repeat myself, when we implement this, we need to keep the k8s specifics pluggable.

Reply:
injecting secrets will be a scaling issue, and in many cases when secrets are created by the plugins (see cephfs), they are not injectable.

Reply:
> I have a few follow-up questions,
> 
>     * are we planning to support ceph-csi driver for multiple CO?

In my discussion with @rootfs we are currently focusing on k8s only. Should provide answer for the rest of the questions.